spacy_experimental/coref/coref_util.py

```diff
@@ -0,0 +1,248 @@
+        # Check antecedent's cluster
+        antecedent = (int(span_starts[predicted_idx]), int(span_ends[predicted_idx]))
+        antecedent_cluster_id = mention_to_cluster_id.get(antecedent, -1)
+        if antecedent_cluster_id == -1:
+            antecedent_cluster_id = len(predicted_clusters)
+            predicted_clusters.append([antecedent])
+            mention_to_cluster_id[antecedent] = antecedent_cluster_id
+        # Add mention to cluster
+        mention = (int(span_starts[i]), int(span_ends[i]))
+        predicted_clusters[antecedent_cluster_id].append(mention)
+        mention_to_cluster_id[mention] = antecedent_cluster_id
+
+    predicted_clusters = [tuple(c) for c in predicted_clusters]
+    return predicted_clusters
+
+
+def select_non_crossing_spans(
+    idxs: List[int], starts: List[int], ends: List[int], limit: int
+) -> List[int]:
+    """Given a list of spans sorted in descending order, return the indexes of
+    spans to keep, discarding spans that cross.
+
+    Nested spans are allowed.
+    """
+    # ported from Model._extract_top_spans
+    selected: List[int] = []
+    start_to_max_end: Dict[int, int] = {}
+    end_to_min_start: Dict[int, int] = {}
+
+    for idx in idxs:
+        if len(selected) >= limit or idx > len(starts):
+            break
+
+        start, end = starts[idx], ends[idx]
+        cross = False
+
+        for ti in range(start, end):
+            max_end = start_to_max_end.get(ti, -1)
+            if ti > start and max_end > end:
+                cross = True
+                break
+
+            min_start = end_to_min_start.get(ti, -1)
+            if ti < end and 0 <= min_start < start:
+                cross = True
+                break
+
+        if not cross:
+            # this index will be kept
+            # record it so we can exclude anything that crosses it
+            selected.append(idx)
+            max_end = start_to_max_end.get(start, -1)
+            if end > max_end:
+                start_to_max_end[start] = end
+            min_start = end_to_min_start.get(end, -1)
+            if min_start == -1 or start < min_start:
+                end_to_min_start[end] = start
+
+    # sort idxs by order in doc
+    selected = sorted(selected, key=lambda idx: (starts[idx], ends[idx]))
+    # This was causing many repetitive entities in the output - removed for now
+    # while len(selected) < limit:
+    #     selected.append(selected[0])  # this seems a bit weird?
+    return selected
+
+
+def create_head_span_idxs(ops, doclen: int):
+    """Helper function to create single-token span indices."""
+    aa = ops.xp.arange(0, doclen)
+    bb = ops.xp.arange(0, doclen) + 1
+    return ops.asarray2i([aa, bb]).T
+
+
+def get_clusters_from_doc(doc) -> List[List[Tuple[int, int]]]:
+    """Convert the span clusters in a Doc to simple integer tuple lists. The
+    ints are char spans, to be tokenization independent.
+    """
+    out = []
+    keys = sorted(list(doc.spans.keys()))
```

Comment:
This will get any other spans that do not belong to coreference clusters, but are maybe assigned by `spancat` or something else. It used to take `prefix` as argument to prevent this. 

Also the name is a bit misleading I guess because it picks the `root` of each span it encounters not the span itself. So its quite specfic, maybe just `get_head_clusters` ?

Reply:
Those are good points. This function is primarily used in testing, where the goal is to check that spans are equivalent, so it should actually be fine to have it work on not just heads, but whole spans.

The one place it's used outside of tests is in loss calculation. I'll add prefix filtering so other spans don't get mixed in there.

Reply:
OK, made these changes. By default this now uses whole spans instead of tokens and no prefix, which is fine for all existing tests. In the `get_loss` of the clusterer, it uses heads only and a prefix filter.